// tsd-jsdoc/dist/doclet_utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function isClassDoclet(doclet) {
    return doclet.kind === 'class' || doclet.kind === 'interface' || doclet.kind === 'mixin';
}
exports.isClassDoclet = isClassDoclet;
function isFileDoclet(doclet) {
    return doclet.kind === 'file';
}
exports.isFileDoclet = isFileDoclet;
function isEventDoclet(doclet) {
    return doclet.kind === 'event';
}
exports.isEventDoclet = isEventDoclet;
function isFunctionDoclet(doclet) {
    return doclet.kind === 'function' || doclet.kind === 'callback';
}
exports.isFunctionDoclet = isFunctionDoclet;
function isMemberDoclet(doclet) {
    return doclet.kind === 'member' || doclet.kind === 'constant';
}
exports.isMemberDoclet = isMemberDoclet;
function isNamespaceDoclet(doclet) {
    return doclet.kind === 'module' || doclet.kind === 'namespace';
}
exports.isNamespaceDoclet = isNamespaceDoclet;
function isTypedefDoclet(doclet) {
    return doclet.kind === 'typedef';
}
exports.isTypedefDoclet = isTypedefDoclet;
function isPackageDoclet(doclet) {
    return doclet.kind === 'package';
}
exports.isPackageDoclet = isPackageDoclet;
function isEnumDoclet(doclet) {
    return isMemberDoclet(doclet) && doclet.isEnum;
}
exports.isEnumDoclet = isEnumDoclet;
